tensorflow/core/kernels/sparse_tensor_dense_matmul_op_gpu.cu.cc

```diff
@@ -71,60 +101,111 @@ __global__ void SparseTensorDenseMatMulKernel(
+    return deterministic_ops;
+  }();
+  return require_determinism;
+}
+
+} // namespace
+
 template <typename T, typename Tindices, bool ADJ_A, bool ADJ_B>
 struct SparseTensorDenseMatMulFunctor<GPUDevice, T, Tindices, ADJ_A, ADJ_B> {
+
+  template <typename Tsum>
   static EIGEN_ALWAYS_INLINE Status
-  Compute(OpKernelContext* ctx, typename TTypes<T>::Matrix out,
-          typename TTypes<Tindices>::ConstMatrix a_indices,
-          typename TTypes<T>::ConstVec a_values,
-          typename TTypes<T>::ConstMatrix b) {
+  GpuSparseTensorDenseMatMulWithRecast(
+      OpKernelContext* ctx, typename TTypes<T>::Matrix &out,
+      typename TTypes<Tindices>::ConstMatrix a_indices,
+      typename TTypes<T>::ConstVec a_values,
+      typename TTypes<T>::ConstMatrix b) {
+
     int nnz = a_values.size();
     // out = A * B, A is [m x n] and B is [n x p], out is [m x p]
     int m = out.dimension(0);
     int p = out.dimension(1);
     int b_rows = b.dimension(0);
     int b_cols = b.dimension(1);
-
     const GPUDevice& d = ctx->eigen_device<GPUDevice>();
-    using Tsum = typename SumType<T>::type;
-    Tsum* maybe_temp_out_data = nullptr;
+    GpuLaunchConfig config = GetGpuLaunchConfig(p * nnz, d);
+
     Tensor temp_out_t;
+    Tsum* maybe_temp_out_data = nullptr;
     bool sum_type_is_different = !std::is_same<T, Tsum>::value;
+
     if (sum_type_is_different) {
       TF_RETURN_IF_ERROR(ctx->allocate_temp(
           DataTypeToEnum<Tsum>::value,
           TensorShape({out.dimension(0), out.dimension(1)}), &temp_out_t));
+
       auto temp_out = temp_out_t.matrix<Tsum>();
       maybe_temp_out_data = temp_out.data();
-      temp_out.device(d) = temp_out.constant(Tsum(0));
     } else {
       // Note: The reinterpret cast is only required to avoid a compilation
       // error; it is only used if Tsum == T.
       maybe_temp_out_data = reinterpret_cast<Tsum*>(out.data());
-      out.device(d) = out.constant(T(0));
     }
 
+    TF_CHECK_OK(GpuLaunchKernel(
+        SetZero<Tsum>, config.block_count, config.thread_per_block, 0,
+        d.stream(), m * p, maybe_temp_out_data));
+
     // TODO(ebrevdo): Should this be alpha * nnz instead of
     // out.size()?  Perhaps p * nnz ?
-    GpuLaunchConfig config = GetGpuLaunchConfig(p * nnz, d);
 
     TF_CHECK_OK(GpuLaunchKernel(
         SparseTensorDenseMatMulKernel<T, Tsum, Tindices, ADJ_A, ADJ_B>,
         config.block_count, config.thread_per_block, 0, d.stream(), nnz, m,
         b_rows, b_cols, p, a_indices.data(), a_values.data(), b.data(),
         maybe_temp_out_data));
 
-    if (sum_type_is_different) {
+    if (RequireDeterminism()) {
+      TF_CHECK_OK(GpuLaunchKernel(
+          DownCast<Tsum, T>, config.block_count, config.thread_per_block,
+          0, d.stream(), m * p, maybe_temp_out_data,
+          out.data()));
+    }
+    else if (sum_type_is_different) {
       out.device(d) = temp_out_t.matrix<Tsum>().template cast<T>();
     }
```

Comment:
What stops us from replacing this block of code (lines 176-182) with the following, simpler (and seemingly faster), code?:

```
if (sum_type_is_different) {
  TF_CHECK_OK(GpuLaunchKernel(
      DownCast<Tsum, T>, config.block_count, config.thread_per_block,
      0, d.stream(), m * p, maybe_temp_out_data,
      out.data()));
}
```

Reply:
It's totally ok to replace, just want to signal there could be two scenarios.

Reply:
Thanks.

`sum_type_is_different == True` may be due to requiring determinism or accumulating float16 operands; that is decided and documented elsewhere. All this pieces of code needs to know, to do what it needs to do, is whether the types are different. The previous code introduced two different code-paths (one probably slower), that each did the same thing in a different way. It added complexity unnecessarily. If you wanted to remind the reader of what could make `sum_type_is_different` be `True`, then a comment would suffice:

`// Note that this branch will be taken either when the user wants determinism or the data type is float16`

However, I don't think it's necessary because this code simply performs the cast operation whenever it's needed. Other/earlier code specifies when that's needed.